kliff/trainer/base_trainer.py

```diff
@@ -0,0 +1,636 @@
+import hashlib
+import importlib
+import json
+import os
+import random
+from copy import deepcopy
+from datetime import datetime, timedelta
+from glob import glob
+from pathlib import Path
+from typing import Callable, Union
+
+import dill  # TODO: include dill in requirements.txt
+import numpy as np
+import yaml
+from loguru import logger
+
+from kliff.dataset import Dataset
+
+
+class Trainer:
+    """Base class for all trainers.
+
+    This class is the base class for all trainers. It provides the basic structure for
+    training a model. The derived classes should implement the required methods. This
+    class will provide the basic functionality for training, such as setting up the
+    work directory, saving the configuration, and setting up the indices for training
+    and validation datasets. It will save hashes of the configuration fingerprints and
+    training configuration to the work directory. This would ensure reproducibility of
+    the training process, and easy restarting.
+
+    The core trainer class will provide the following functionality:
+    - Set up the work directory
+    - Set up the dataset
+    - Set up the test train split
+    Model, parameter transform and optimizer setup are left for the derived classes to
+    implement.
+
+    Args:
+        training_manifest: training manifest
+    """
+
+    def __init__(self, training_manifest: dict, model=None):
+        # workspace variables
+        self.workspace: dict = {
+            "name": "kliff_workspace",
+            "seed": 12345,
+            "resume": False,
+            "walltime": "2:00:00:00",
+        }
+
+        # dataset variables
+        self.dataset_manifest: dict = {
+            "type": "kliff",
+            "path": "./",
+            "save": False,
+            "shuffle": False,
+            "keys": {"energy": "energy", "forces": "forces"},
+            "colabfit_dataset": {
+                "dataset_name": None,
+                "database_name": None,
+                "database_url": None,
+            },
+        }
+        self.dataset = None
+
+        # model variables
+        self.model_manifest: dict = {
+            "type": "kim",
+            "name": None,
+            "path": None,
+        }
+        self.model: Callable = model
+
+        # transform variables
+        self.transform_manifest: dict = {
+            "property": [
+                {
+                    "name": None,
+                    "property_key": None,
+                }
+            ],
+            "parameter": [],
+            "configuration": {
+                "name": None,
+                "kwargs": None,
+            },
+        }
+        self.property_transforms = []
+        self.configuration_transform = None
+
+        # training variables
+        # this is too complicated to put it in singe dict, therefore the training
+        # block is divided into loss, optimizer, dataset_sample
+        self.training_manifest: dict = {}
```

Comment:
I dont understand, parse manifest divides the `training_manifest` in loss optimizer etc. Also training manifest also contains information that cannot be categorized so neatly, so it can be used. It is just that loss_manifest etc provides easy abstractions to categorize these data, as opposed to a single monolith dictionary. For example, exponential averaging of weights, can be categorized as model, or optimizer, or training manifest, so for such information `training_manifest` can be used.